File: 2023/day10/solution.py
#!/usr/bin/env python3
import sys
from functools import cache
from dataclasses import dataclass
from typing import Self
from collections import deque

# yapf: disable
sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

@dataclass
class Point():
  r: int; c: int
  def __mul__(self, value: int): return Point(self.r * value, self.c * value)
  def __add__(self, other: Self): return Point(self.r + other.r, self.c + other.c)
  def __lt__(self, other: Self): return (self.r, self.c) < (other.r, other.c)
  def __hash__(self): return hash((self.r, self.c))

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i': FILENAME = 'input.txt'
      case '-p1': PART1, PART2 = True, False
      case '-p2': PART1, PART2 = False, True
      case '-p0': PART1, PART2 = False, False
      case _: raise Exception(f'unknown {arg=}')

D4 = [Point(*x) for x in [(-1, 0), (0, 1), (1, 0), (0, -1)]]

def debug(level: int, *args, **kwargs):
  if level < DEBUG: print(*args, **kwargs)
#yapf: enable

TILE_ENTRY = {
  'S': [0, 1, 2, 3],
  '|': [0, 2],
  '-': [1, 3],
  'F': [0, 3],
  '7': [0, 1],
  'L': [2, 3],
  'J': [2, 1],
}
TILE_EXIT = {k: [(x + 2) % 4 for x in v] for k, v in TILE_ENTRY.items()}

with open(FILENAME, 'r') as f:
  text = f.read()
grid = text.splitlines()
rows, cols = len(grid), len(grid[0])


def find_grid(find_char: str):
  for r, row in enumerate(grid):
    for c, char in enumerate(row):
      if char == find_char: return Point(r, c)
  assert False


def find_path(start: Point) -> set[Point]:
  walked = set([start])
  dir_walked = set([(start.r, start.c, -1)])  # only for visualisation
  queue = deque([start])
  while queue:
    current = queue.pop()
    current_char = grid[current.r][current.c]
    for d in TILE_EXIT[current_char]:
      offset = D4[d]
      next_tile = current + offset
      if not (0 < next_tile.r < rows and 0 < next_tile.c < cols): continue
      if next_tile in walked: continue
      next_char = grid[next_tile.r][next_tile.c]
      debug(1, f'{current=} {current_char=} {d=} {next_char=}')
      if d not in TILE_ENTRY.get(next_char, []):
        debug(2, '  not in tile entry')
        continue
      queue.append(next_tile)
      walked.add(next_tile)
      dir_walked.add((next_tile.r, next_tile.c, d))
  loop_length = len(walked)
  if DEBUG > 1:
    grid_copy = [list(x) for x in grid]
    for r, c, d in dir_walked:
      content = grid_copy[r][c]
      grid_copy[r][c] = f'\x1b[7;{32+d}m{content}\x1b[0m'
    for r, row in enumerate(grid_copy):
      print(f'''{r:3d}: {''.join(row)}''')
  return walked


def part1():
  # a story about some chud who beleived it was their right to harass animals :|
  tiles = find_path(find_grid('S'))
  print(f'part 1: {len(tiles)//2+1}')
  # 6757


def part2():
  # there's a walk solution to this but it's non-trivial

  start = find_grid('S')
  path = find_path(start)
  # evaluate tiles surrounding S. check if they're in path and S was allowed to enter them
  start_dirs: list[int] = []
  for d, offset in enumerate(D4):
    next_tile = start + offset
    if next_tile in path and d in TILE_ENTRY[grid[next_tile.r][next_tile.c]]:
      start_dirs.append(d)
  # inferred start is a reverse lookup from TILE_EXIT
  inferred_start = {tuple(sorted(v)): k for k, v in TILE_EXIT.items()}.get(tuple(sorted(start_dirs)))
  debug(0, f'{start=} {start_dirs=} {inferred_start=}')
  row_list = list(grid[start.r])
  row_list[start.c] = inferred_start
  # update the grid so S doesn't throw off our enclosed scan
  grid[start.r] = row_list

  enclosed: set[Point] = set()
  for r, row in enumerate(grid):
    is_enclosed = False
    for c, char in enumerate(row):
      # for path tiles, flip is_enclosed when the tile has an up component. down also works as long as we're consistent
      if Point(r, c) in path:
        if 0 in TILE_ENTRY[char]: is_enclosed ^= True
      # add is_enclosed non-path tiles to the enclosed set
      elif is_enclosed:
        enclosed.add(Point(r, c))
  if DEBUG > 1:
    grid_copy = [list(x) for x in grid]
    for tile in enclosed:
      content = grid_copy[tile.r][tile.c]
      grid_copy[tile.r][tile.c] = f'\x1b[7;31m{content}\x1b[0m'
    for r, row_list in enumerate(grid_copy):
      print(f'''{r:3d}: {''.join(row_list)}''')
  print(f'part 2: {len(enclosed)}')
  # 523


if PART1: part1()
if PART2: part2()

File: 2023/day10/solution.ts
#!/usr/bin/env -S deno --allow-read

const Maths = Math;

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

class HashedSet<K, H> {
  // hopefully not too slow but we'll see
  private map = new Map<H, K>();
  private hashFn: (key: K) => H;
  constructor(hashFn: (key: K) => H, iterable?: Iterable<K>) {
    this.hashFn = hashFn;
    if (iterable) {
      for (const k of iterable) {
        this.map.set(this.hashFn(k), k);
      }
    }
  }
  add(key: K): void {
    this.map.set(this.hashFn(key), key);
  }
  has(key: K): boolean {
    return this.map.has(this.hashFn(key));
  }
  values(): MapIterator<K> {
    return this.map.values();
  }
  size(): number {
    return this.map.size;
  }
}

class Tile {
  constructor(public r: number, public c: number) {}
  add(other: Tile): Tile {
    return new Tile(this.r + other.r, this.c + other.c);
  }
}

const D4 = [new Tile(-1, 0), new Tile(0, 1), new Tile(1, 0), new Tile(0, -1)];
const ENTRY = new Map<string, number[]>([
  ['|', [0, 2]],
  ['-', [1, 3]],
  ['L', [2, 3]],
  ['J', [2, 1]],
  ['7', [0, 1]],
  ['F', [0, 3]],
  ['S', [0, 1, 2, 3]],
]);
const EXIT = new Map<string, number[]>(ENTRY.entries().map(([k, v]) => [k, v.map((e) => (e + 2) % 4)]));

const input = await Deno.readTextFile(FILENAME);

const parseInput = () => {
  const grid = input.split('\n').filter((line) => line.trim());
  const rows = grid.length;
  const cols = grid[0].length;
  const gridFind = (search: string) => {
    for (const [r, row] of grid.entries()) {
      for (const [c, char] of row.split('').entries()) {
        if (char == search) return new Tile(r, c);
      }
    }
    throw new Error('bruh');
  };
  return { grid, rows, cols, start: gridFind('S') };
};

const findPath = (grid: string[], rows: number, cols: number, start: Tile) => {
  // standard aoc path walk with some additional direction constraints
  const path = new HashedSet<Tile, string>((key) => `${key.r},${key.c}`);
  path.add(start);
  const queue: Tile[] = [start];
  while (queue.length) {
    const tile = queue.shift()!;
    for (const direction of EXIT.get(grid[tile.r][tile.c])!) {
      const nextTile = tile.add(D4[direction]);
      if (nextTile.r < 0 || nextTile.r >= rows || nextTile.c < 0 || nextTile.c >= cols || path.has(nextTile)) continue;
      if (!(ENTRY.get(grid[nextTile.r][nextTile.c]) ?? []).includes(direction)) continue;
      path.add(nextTile);
      queue.push(nextTile);
    }
  }
  return path;
};

const part1 = () => {
  const { grid, rows, cols, start } = parseInput();
  debug(1, { grid, rows, cols, start });
  const path = findPath(grid, rows, cols, start);
  debug(1, { path });
  console.log('part 1:', Maths.floor(path.size() / 2));
};

const part2 = () => {
  const { grid, rows, cols, start } = parseInput();
  const path = findPath(grid, rows, cols, start);
  // replace S so it won't interfere with our row scan
  const startDirections: number[] = [];
  for (const [direction, offset] of D4.entries()) {
    const nextTile = start.add(offset);
    if (
      nextTile.r < 0 ||
      nextTile.c >= rows ||
      nextTile.c < 0 ||
      nextTile.c >= cols ||
      !path.has(nextTile) ||
      !ENTRY.get(grid[nextTile.r][nextTile.c])!.includes(direction)
    )
      continue;
    startDirections.push(direction);
  }
  const inferred = new Map(EXIT.entries().map(([k, v]) => [v.toSorted().join(','), k])).get(startDirections.join(','))!;
  debug(1, { startDirections, inferred });
  // grid scan, flip isInternal on path tiles with an up direction. add isInternal non-path tiles to set (map actually but shush)
  grid[start.r] = grid[start.r]
    .split('')
    .map((item, i) => (i == start.c ? inferred : item))
    .join('');
  const internal = new HashedSet<Tile, string>((key) => `${key.r},${key.c}`);
  for (const [r, row] of grid.entries()) {
    let isInternal = false;
    for (const [c, char] of row.split('').entries()) {
      const tile = new Tile(r, c);
      debug(2, { char, tile });
      if (path.has(tile)) {
        if (ENTRY.get(char)!.includes(0)) isInternal = !isInternal;
      } else if (isInternal) {
        internal.add(tile);
      }
    }
  }
  debug(1, { internal });
  console.log('part 2:', internal.size());
};

if (PART1) part1();
if (PART2) part2();
